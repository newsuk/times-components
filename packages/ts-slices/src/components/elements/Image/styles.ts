import styled from 'styled-components';
import { breakpoints } from '@times-components/ts-styleguide';

import { DisplaySchema } from '../../../types/styles';
import { isSideBySide, isImageHidden } from '../../../utils/getArticleStyles';

const getMargin = (breakpoint: string, schema?: DisplaySchema) =>
  isSideBySide(breakpoint, schema) || isImageHidden(breakpoint, schema)
    ? 0
    : '12px';

const getDisplay = (breakpoint: string, schema?: string[]) =>
  schema && schema.includes(breakpoint) ? 'block' : 'none';

const getImagePadding = (ratio?: string) => {
  switch (ratio) {
    case '16:9':
      return '56.25%';
    case '3:2':
      return '66.66%';
    case '1:1':
      return '100%';

    default:
      return 0;
  }
};

export const ImageContainer = styled.div<{ schema?: DisplaySchema }>`
  margin-bottom: ${({ schema }) => getMargin('sm', schema)};

  @media (min-width: ${breakpoints.medium}px) {
    margin-bottom: ${({ schema }) => getMargin('md', schema)};
  }

  @media (min-width: ${breakpoints.wide}px) {
    margin-bottom: ${({ schema }) => getMargin('lg', schema)};
  }

  @media (min-width: ${breakpoints.huge}px) {
    margin-bottom: ${({ schema }) => getMargin('xlg', schema)};
  }
`;

export const ImageRatio = styled.div<{ ratio: string; schema?: string[] }>`
  position: relative;
  display: ${({ schema }) => getDisplay('sm', schema)};
  overflow: hidden;

  &:after {
    content: '';
    display: block;
    padding-bottom: ${({ ratio }) => getImagePadding(ratio)};
  }

  @media (min-width: ${breakpoints.medium}px) {
    display: ${({ schema }) => getDisplay('md', schema)};
  }

  @media (min-width: ${breakpoints.wide}px) {
    display: ${({ schema }) => getDisplay('lg', schema)};
  }

  @media (min-width: ${breakpoints.huge}px) {
    display: ${({ schema }) => getDisplay('xlg', schema)};
  }
`;
