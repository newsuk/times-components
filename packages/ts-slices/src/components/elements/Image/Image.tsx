import React from 'react';
import get from 'lodash.get';

import { SliceArticle } from '../../../types/slice';
import { MouseEventType, ClickHandlerType } from '../../../types/event';
import { DisplaySchema } from '../../../types/styles';
import { getImageRatioSchema } from '../../../utils/getArticleStyles';
import { getImageByRatio, getImageAltText } from '../../../utils/getImage';

import { LazyImage } from './LazyImage/LazyImage';

import { ImageContainer, ImageRatio } from './styles';

export const Image: React.FC<{
  article: SliceArticle;
  displaySchema?: DisplaySchema;
  clickHandler?: ClickHandlerType;
}> = ({ article, displaySchema, clickHandler }) => {
  const ratioSchema = getImageRatioSchema(displaySchema);

  const onClick = (event: MouseEventType) => {
    if (article && clickHandler) {
      clickHandler(event, article);
    }
  };

  return (
    <ImageContainer schema={displaySchema}>
      <a onClick={onClick} href={article.url}>
        {Object.entries(ratioSchema).map(([ratio, schema], i: number) => {
          const image = getImageByRatio(ratio, article);

          return (
            <ImageRatio key={i} ratio={ratio} schema={schema}>
              <LazyImage
                url={get(image, 'url')}
                alt={getImageAltText(article)}
                isRoundal={ratio === '1:1'}
                isBackground={get(image, 'ratio') === '*'}
              />
            </ImageRatio>
          );
        })}
      </a>
    </ImageContainer>
  );
};
