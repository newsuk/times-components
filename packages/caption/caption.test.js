/* eslint-env jest */

import "react-native";
import React from "react";
import renderer from "react-test-renderer";
import Image from "@times-components/image";
import Caption from "./caption";

const captionText = "Some caption text goes in here";
const credits = "Just credits";
const exampleImage = {
  uri:
    "https://www.thetimes.co.uk/imageserver/image/methode%2Fsundaytimes%2Fprod%2Fweb%2Fbin%2F9242e576-4dfc-11e7-a20e-a11097d3353d.jpg?crop=1463%2C975%2C293%2C12&resize=320"
};
const style = {
  container: {
    backgroundColor: "red"
  },
  text: {
    color: "green"
  }
};

it("renders correctly without credits", () => {
  const tree = renderer.create(<Caption text={captionText} />).toJSON();

  expect(tree).toMatchSnapshot();
});

it("renders correctly with credits", () => {
  const tree = renderer
    .create(<Caption text={captionText} credits={credits} />)
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it("renders correctly with credits only", () => {
  const tree = renderer.create(<Caption credits={credits} />).toJSON();

  expect(tree).toMatchSnapshot();
});

it("renders correctly with specific styles", () => {
  const tree = renderer
    .create(<Caption text={captionText} credits={credits} style={style} />)
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it("renders an image with a caption", () => {
  const tree = renderer
    .create(
      <Caption text={captionText} credits={credits}>
        <Image source={exampleImage} />
      </Caption>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
