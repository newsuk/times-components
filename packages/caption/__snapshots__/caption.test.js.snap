// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with credits 1`] = `
<View
  style={
    Object {
      "paddingTop": 10,
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "color": "#696969",
        "fontFamily": "GillSansMTStd-Medium",
        "fontSize": 13,
        "lineHeight": 15,
      }
    }
  >
     
    Some caption text goes in here
     
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Array [
        Object {
          "color": "#696969",
          "fontFamily": "GillSansMTStd-Medium",
          "fontSize": 13,
          "lineHeight": 15,
        },
        Object {
          "color": "#333",
          "fontSize": 9,
          "fontWeight": "400",
          "letterSpacing": 1,
          "marginTop": 2,
        },
      ]
    }
  >
    JUST CREDITS
  </Text>
</View>
`;

exports[`renders correctly without credits 1`] = `
<View
  style={
    Object {
      "paddingTop": 10,
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "color": "#696969",
        "fontFamily": "GillSansMTStd-Medium",
        "fontSize": 13,
        "lineHeight": 15,
      }
    }
  >
     
    Some caption text goes in here
     
  </Text>
  
</View>
`;
